python/rmm/_lib/memory_resource.pyx

```diff
@@ -56,9 +57,24 @@ cdef extern from "rmm/mr/device/managed_memory_resource.hpp" \
 
 cdef extern from "rmm/mr/device/cuda_async_memory_resource.hpp" \
         namespace "rmm::mr" nogil:
+
     cdef cppclass cuda_async_memory_resource(device_memory_resource):
-        cuda_async_memory_resource(optional[size_t] initial_pool_size,
-                                   optional[size_t] release_threshold) except +
+        cuda_async_memory_resource(
+            optional[size_t] initial_pool_size,
+            optional[size_t] release_threshold,
+            optional[allocation_handle_type] export_handle_type) except +
+
+# TODO: when we adopt Cython 3.0 use enum class
+cdef extern from "rmm/mr/device/cuda_async_memory_resource.hpp" \
+        namespace \
+        "rmm::mr::cuda_async_memory_resource::allocation_handle_type" \
+        nogil:
+    enum allocation_handle_type \
+            "rmm::mr::cuda_async_memory_resource::allocation_handle_type":
+        none
+        posix_file_descriptor
+        win32
+        win32_kmt
```

Comment:
I have only limited familiarity with this, but the [patterns from cudf](https://github.com/rapidsai/cudf/blob/branch-22.06/python/cudf/cudf/_lib/cpp/types.pxd) and my prior knowledge from conversations with @shwina would indicate that this needs C names to avoid mangling. If not, why is this different than cuDF?
```suggestion
    ctypedef enum allocation_handle_type \
            "rmm::mr::cuda_async_memory_resource::allocation_handle_type":
        none "rmm::mr::cuda_async_memory_resource::allocation_handle_type::none"
        posix_file_descriptor "rmm::mr::cuda_async_memory_resource::allocation_handle_type::posix_file_descriptor"
        win32 "rmm::mr::cuda_async_memory_resource::allocation_handle_type::win32"
        win32_kmt "rmm::mr::cuda_async_memory_resource::allocation_handle_type::win32_kmt"
```

Reply:
@shwina is being sneaky (and kind of cheating) :) If you look at the `cdef extern` statement, you'll notice that he's setting the namespace to `rmm::mr::cuda_async_memory_resource::allocation_handle_type`, not `**rmm::mr::cuda_async_memory_resource`. That implies that _everything_ contained in that block will be within that namespace. In this particular instance that happens to be true, so this works. In the example @bdice linked to, though, we are getting a bunch of different types from different namespaces in cudf, so we need to namespace each member.

The problem is that C enums are not namespaced, so the namespace applied to the enum itself doesn't apply to its members, only to the enum itself, i.e. anywhere that you actually write `enum X` in the code it will have the correct namespace, but not when you access its members. Every enum inside the `cdef extern` block gets the namespace of the block, unless you take advantage of the usual trick we use (as in @bdice's link) which is to provide a complete textual replacement for that token. I don't think the ctypedef matters in most cases for enums since Cython will take care of adding the "enum" in the generated code as needed.

Reply:
I guess this is fine, then. In my understanding, `enum class` support in Cython 3 should also solve this because its members should be namespaced correctly?

I'm also fine with dropping the `ctypedef`. In general it would be nice to avoid "tricks" and for RAPIDS Cython code to look and feel syntactically similar - it helps assess correctness when new code looks like existing (trusted) code.

Reply:
Yes, Cython 3 properly supports C++ scoped enums. I agree that making the code look similar to familiar code is helpful for less experienced readers and would be fine making it look identical (including the `ctypedef`) for that purpose.

Reply:
> In general it would be nice to avoid "tricks" and for RAPIDS Cython code to look and feel syntactically similar

I think with Cython 3 this will be a little easier. Unfortunately, tricks are the only way we can really support enum classes until then. Adding a C-name specification to each individual member becomes much harder to read with all the line-continuation we'd need to do.

Reply:
Yeah I guess to be clear, this whole thing is a "trick". Some tricks are only better than other tricks insofar as they are already familiar. I'm going to close this thread as resolved since it's not clear that making the proposed change is really an improvement.